File: 98-validate-binary-search-tree/98-validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def isValidBST(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        stack = []
        #inorder traversal
        cur = root
        res = []
        while(cur or stack):
            while(cur):
                stack.append(cur)
                cur = cur.left
            popped = stack.pop()
            if popped and res and res[-1].val >= popped.val:
                return False
            res.append(popped)
            cur = popped.right
        return True

File: 98-validate-binary-search-tree/98-validate-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isValidBST(TreeNode root) {
        int prev = Integer.MIN_VALUE;
        LinkedList<TreeNode> stack = new LinkedList<>();
        TreeNode cur = root;
        int count = 0;
        while(!stack.isEmpty() || cur != null){
            while(cur!=null){
                stack.addLast(cur);
                cur = cur.left;
            }
            TreeNode popped = stack.pollLast();
            if(popped.val < prev)return false;
            if(prev == popped.val && prev != Integer.MIN_VALUE)return false;
            if(prev == popped.val && prev == Integer.MIN_VALUE && count > 0)return false;
            prev = popped.val;
            cur = popped.right;
            count += 1;
            
        }
        return true;
        
        
    }
}